Add a layer to show where on the map problems are occurring, either i…

dabreegster opened the pull request:
…ndividually or as a heatmap. #600

This was super quick to crank out by just copying what we've already got for the population map. It also seems pretty decent at quickly pinpointing actual problematic areas:
![Screenshot from 2021-04-22 14-44-38](https://user-images.githubusercontent.com/1664407/115789361-e26b8680-a379-11eb-80cc-d07269d6b05a.png)

Video demo (gif too big, still having trouble encoding to mp4 and github accepting it): https://www.dropbox.com/s/ctjzh1073qb5o4m/problem_demo.gif?dl=0

dabreegster commented on game/src/layer/problems.rs:
```diff
@@ -0,0 +1,191 @@
+            ))
+        } else {
+            let circle = Circle::new(Pt2D::new(0.0, 0.0), Distance::meters(10.0)).to_polygon();
+            // TODO Different colors per problem type
```
Or at the very least, not using red, because that looks like unzoomed cars

dabreegster commented on game/src/layer/problems.rs:
```diff
@@ -0,0 +1,191 @@
+impl Options {
+    pub fn new() -> Options {
+        Options {
+            heatmap: None,
```
I thought the heatmap would be slow to calculate, but actually it's not, and it seems easier to spot major areas faster. Opinions?

michaelkirk replied:
Seems reasonable to me. 👍

dabreegster commented:
Also asking for @Robinlovelace's thoughts on this dataviz, as it's usable for historic collision data, and I'm sure you've encountered plenty of tools that show things like this before. Async review fine; fast iteration, not long reviews, are my vibe

dabreegster commented:
Oh yeah, and the layer isn't intended to show _change_ in problems due to edits, just the current state of things. And I'll add a time filter in the future, to quickly see if a problem affects an area only part of the day.

michaelkirk reviewed (approved):
When I first enabled this feature, I was confused because I didn't see anything happen on the map.

I was zoomed in, and apparently when zoomed in we only decorate the minimap, but when there are already a bunch of agent dots on the minimap, its hard to notice a change if you don't already know where to expect it (this might be exacerbated by our moving the layer controls farther away from the minimap.)

Do you think it would make sense to show some kind of icon on the screen while zoomed in when this layer is present? Ideally something distinct per problem and/or with a distinct tooltip on hover.

Aside from that, when zoomed out, everything looks pretty good. It might be nice to color-code or otherwise stylize the different problem markers at that zoom level too, but... between all the agent colors, we might be asking too much to do it with color alone.

Overall, LGTM, I think it's useful as is, and nothing stops us from changing it in the future.

dabreegster commented:
I switched to show the heatmap by default and use purple, not red dots.

Also added a time slider filter, even though actually using the sliders is clunky -- no smoothly dragging, only clicking to set a new state. Will keep iterating going forward.

dabreegster commented:
> Do you think it would make sense to show some kind of icon on the screen while zoomed in when this layer is present? Ideally something distinct per problem and/or with a distinct tooltip on hover.

Yuwen actually brought this up after the latest UX study. Most layers don't show anything zoomed in, only unzoomed or on the minimap. If we decide this is actually a good thing and not just kind of accidental, then we may try automatically unzooming when you open a layer while zoomed in. I'm not sure if it'll be distracting / unexpected, or cause problems for people who want to return the camera to where it was.

But to your point here, we _could_ add more detail and allow tooltips here. The implementation will need to aggregate problems per point, since everything right now maps to the center of a lane or intersection. The tooltip could show the count broken down by problem type and _maybe_ even some histogram of when the problems occurred through time, and even give you a way to time-warp and sample a few of the problems.

Robinlovelace commented:
Will take a look. I'm just rebuilding the latest version on this branch to give it a go with

```
git checkout safety_layer         
Branch 'safety_layer' set up to track remote branch 'safety_layer' from 'origin'.
Switched to a new branch 'safety_layer'

abstreet on  safety_layer [?] via 𝗥 v1.50.0 
➜ git pull                 
Already up-to-date.

abstreet on  safety_layer [?] via 𝗥 v1.50.0 took 3s 
➜ cargo build 
```

Will give it a go later today at a time when we're both awake (probably after 5pm UK time).
